### PR TITLE
Service page fix

### PR DESCRIPTION
**Pull request by AngryBerryMS**
For service-page, we need to filter items that matches with `TopLevelTOC` and add them to `TopLevelTOC`.
Then write `TopLevelTOC` back to `ReferenceTOC`.

For `TopLevelTOC` is a generated file and will be dropped finally, so all hrefs use relative path to `ReferenceTOC` unless those start with `~`. And we use `ReferenceTOC` to calculate the relative href of service pages.

[AB#299330](https://dev.azure.com/ceapex/Engineering/_boards/board/t/DocFX%20v3%20Team/Stories/?workitem=299330)

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/6551)

### CODE REVIEW

File: docs/specs/toc.yml [hunk 1]
```diff
@@ -1895,3 +1890,52 @@ outputs:
     {"metadata": {"universal_conceptual_toc": "dotnet-test/learn/toc.json"}}
   test/a/_splitted/b/toc.json: |
     {"metadata": {"universal_conceptual_toc": "dotnet-test/learn/toc.json"}}
+---
+# Service Page href fix
```
Comment (live1206) [inline docs/specs/toc.yml:1894]: What is different from the existing test?

File: src/docfx/build/toc/ServicePageGenerator.cs [hunk 2]
```diff
@@ -86,9 +86,10 @@ internal class ServicePageGenerator
                                 hrefFileFullPath = Path.GetFullPath(Path.Combine(topLevelTOCYmlDir == null ? "" : topLevelTOCYmlDir, childHref));
                             }
 
-                            var servicePageFullPath = Path.GetDirectoryName(Path.GetFullPath(Path.Combine(_docsetPath, servicePagePath.Path))) ?? _docsetPath;
-                            var hrefRelativePath = Path.GetRelativePath(servicePageFullPath, hrefFileFullPath);
-                            childHref = hrefRelativePath;
+                            var hrefRelativePathToDocset = Path.GetRelativePath(_docsetPath, hrefFileFullPath);
+                            var hrefFullPathOfDocset = $"~/{hrefRelativePathToDocset}";
```
Comment (live1206) [inline src/docfx/build/toc/ServicePageGenerator.cs:90]: `~` does not always represent to `_docsetPath`

File: src/docfx/build/toc/TableOfContentsLoader.cs [hunk 3]
```diff
@@ -67,13 +67,14 @@ internal class TableOfContentsLoader
                 var servicePages = new List<FilePath>();
                 var referencedFiles = new List<FilePath>();
                 var referencedTocs = new List<FilePath>();
-                var node = LoadTocFile(file, file, referencedFiles, referencedTocs);
+                TableOfContentsNode node = new TableOfContentsNode();
```
Comment (live1206) [inline src/docfx/build/toc/TableOfContentsLoader.cs:70]: initialization is not needed

File: src/docfx/build/toc/ServicePageGenerator.cs (hunk 2 above)
Comment (live1206) [inline src/docfx/build/toc/ServicePageGenerator.cs:90]: what if we do nothing here, and calculate the relative path based on the `original relative path`, `path of top level toc` and `path of reference toc`?

File: docs/specs/toc.yml (hunk 1 above)
Comment (AngryBerryMS) [reply docs/specs/toc.yml]: Add description

File: src/docfx/build/toc/ServicePageGenerator.cs (hunk 2 above)
Comment (AngryBerryMS) [reply src/docfx/build/toc/ServicePageGenerator.cs]: Finished.

Comment (AngryBerryMS) [reply src/docfx/build/toc/ServicePageGenerator.cs]: Fixed.

Comment (AngryBerryMS) [reply src/docfx/build/toc/ServicePageGenerator.cs]: Fixed/

File: src/docfx/build/toc/TableOfContentsLoader.cs (hunk 3 above)
Comment (AngryBerryMS) [reply src/docfx/build/toc/TableOfContentsLoader.cs:70]: For the return statements in L95, it's needed.

Comment (live1206) [reply src/docfx/build/toc/TableOfContentsLoader.cs:70]: It is initialized either in L76 or L92

Comment (AngryBerryMS) [reply src/docfx/build/toc/TableOfContentsLoader.cs]: Fixed.

File: src/docfx/build/toc/TableOfContentsLoader.cs [hunk 4]
```diff
@@ -85,19 +86,56 @@ internal class TableOfContentsLoader
 
                         node.Items = LoadTocNodes(node.Items, file, file, referencedFiles, referencedTocs);
                     }
+                    else
```
Comment (live1206) [inline src/docfx/build/toc/TableOfContentsLoader.cs:89]: This else can be combined with the below one

File: src/docfx/build/toc/TableOfContentsLoader.cs [hunk 5]
```diff
@@ -113,12 +151,12 @@ private void TraverseAndMerge(TableOfContentsNode node, List<SourceInfo<TableOfC
...
             foreach (var item in node.Items)
             {
-                TraverseAndMerge(item, itemsToMatch, matched);
+                TraverseAndMerge(item, itemsToMatch, matched, joinTOCConfig);
             }
         }
 
         private TableOfContentsNode LoadTocFile(
-            FilePath file, FilePath rootPath, List<FilePath> referencedFiles, List<FilePath> referencedTocs)
+            FilePath file, FilePath rootPath, List<FilePath> referencedFiles, List<FilePath> referencedTocs, bool loadTocNodes = false)
```
Comment (live1206) [inline src/docfx/build/toc/TableOfContentsLoader.cs:159]: consider remove the default value, actually the default behavior should be `true`

File: src/docfx/build/toc/TableOfContentsLoader.cs [hunk 6]
```diff
@@ -135,13 +146,32 @@ private void TraverseAndMerge(TableOfContentsNode node, List<SourceInfo<TableOfC
                 t_recursionDetector.Value = recursionDetector;
 
                 var node = _parser.Parse(file, _errors);
+
+                // Generate service pages.
```
Comment (live1206) [inline src/docfx/build/toc/TableOfContentsLoader.cs:150]: consider putting this in a separate method

**Review by live1206 — APPROVED** [review]

**Pull request merged**